ACOUSTIC PINGER: Add pinger perception and mission

ironmig opened the pull request:
From lakeday 11/12-11/13

1.  Fix tf for hydrophones
2. Add pinger finder perception which locates pingers
3. Add mission to go to gates, locate pinger, and go through correct gate

ironmig commented:
BTW, lakeday changes (with dirty commits) are here: https://github.com/uf-mil/Navigator/compare/dev...lake_day_11-13?expand=1. You can fetch the branch checkout the files you changed, look them over, and send a pull request for your changes with proper commit messages.

tessbianchi commented on perception/navigator_vision/nodes/pinger_finder.py:
```diff
@@ -0,0 +1,170 @@
+# Calculates the point in the plane with the least cummulative distance to
+# every line in line_arr
+def LS_intersection(line_arr):
+    global pinger_position
```
This should be a class, global variables are bad practice.

tessbianchi commented on perception/navigator_vision/nodes/pinger_finder.py:
```diff
@@ -0,0 +1,170 @@
+# receives pings and adds the representation of a line between the hydrophone array and the pinger
+# to our line_arr linear equation system
+def ping_cb(processed_ping):
+    global line_array, sample_amplitudes
```
another global

tessbianchi commented on perception/navigator_vision/nodes/pinger_finder.py:
```diff
@@ -0,0 +1,170 @@
+    return {'success': True, 'message': ""}
+
+def set_freq(msg):
+    global target_freq, line_array, sample_amplitudes, pinger_position
```
Another global

tessbianchi commented on perception/navigator_vision/nodes/pinger_finder.py:
```diff
@@ -0,0 +1,170 @@
+    return SetFrequencyResponse()
+
+
+rospy.init_node('pinger_finder')
```
This needs to be put into a main function or else it will be called if something else imports it.

mattlangford replied:
I agree, also all the stuff at the top should be enclosed too, especially that publisher (which probably should go after the init)

ironmig commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+            yield self.navigator.move.set_position(p).go(initial_plan_time=5)
+
+    @txros.util.cancellableInlineCallbacks
+    def set_active_pinger(self):
```
This depends on Gate_1 being between red-white and Gate_3 between white-green. Ideally the database colorer would set the Gates to the correct color. For example, if the gate is between a white_green_buoy, it will be labeled "GateWhiteGreen". Or it could just set the color of the Gate object (RED,WHITE, or GREEN).

mattlangford replied:
The database colorer colors all buoys it sees (or tries to anyway) in the database. You could try that.
It's going to be sketchy in a lot of cases (as most vision things are) when the sun is infront of us. We may have to turn around after we go through and identify the color (or go through backwards which would be neat).

ironmig commented on perception/navigator_vision/nodes/pinger_finder.py:
```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python
```
Let's add this the perception.launch file. Also, is there anything else related to this mission that needs to be launched with master_launch? Is it just the driver and this?

mattlangford replied:
David said this isn't ready to merge. Probably shouldn't add this and let him do it or talk with him.

mattlangford replied:
I'm a bitch. Disregard all previous comments.

ironmig commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+from visualization_msgs.msg import Marker,MarkerArray
+import navigator_tools
+
+class PingerMission:
```
One issue in the future is that there may be multiple acoustic pinger challenges on the course at a time (with different frequencies). There needs to be a way for the mission planner to tell this mission the id's of the gates it is investigating, and the frequency of the pinger to look for.

ironmig commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+            pose2 = self.gate_2_pos
+        elif index == 2:
+            pose2 = self.gate_3_pos
+        if self.negate:
```
This negate flag was added after lake day. It is intended to prevent the boat from going through the gate before it has listened for pingers. It is not tested yet, but I'll test with bags this week.

ironmig commented on perception/navigator_vision/nodes/pinger_finder.py:
```diff
@@ -0,0 +1,170 @@
+    global debug_arrow_id
+    marker = Marker()
+    marker.ns = "pinger/debug"
+    marker.header.stamp = rospy.Time(0)
```
Maybe this should be set to the current time, ```ros::Time()``` . That may be why the markers are sticking around too long.

ironmig commented on perception/navigator_vision/nodes/pinger_finder.py:
```diff
@@ -0,0 +1,170 @@
+            p1 = p0 + point_distance * delta / npl.norm(delta)
+            line_coeffs = np.array([[p0[0], p0[1], p1[0], p1[1]]]) # p0 and p1 define a line
+            visualize_line(Point(p0[0], p0[1], 0), Point(p1[0], p1[1], 0))
+            if listen:
```
Maybe ```if listen``` should be at the top of the callback. Shouldn't this node just idle if listen is false? This would save on the computational load when this is launched at startup. Also would provide more accurate markers.

mattlangford commented on mission_control/navigator_launch/launch/perception.launch:
```diff
@@ -4,4 +4,5 @@
   <!--node name="start_gate_vision" pkg="navigator_controller" type="start_gate_manual.py" respawn="True"/-->
 
   <node name="scan_the_code" pkg="navigator_vision" type="model_detector.py" respawn="True"/>
+  <node name="pinger_finder" pkg="navigator_vision" type="pinger_finder.py" />
```
Can you move the pinger finder script since it's not vision (at all)

DSsoto replied:
This is supposed to go in the sonar package, which is currently a dependency from Sub8.
It is definitely perception but currently has no place in our file structure.

ironmig replied:
Maybe we should just re-re name the package navigator_perception

mattlangford replied:
That's what it used to be, but that's too general, there's lidar perception, computer vision perception, and now sonar perception.

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+        self.markers_pub = self.navigator.nh.advertise("/pinger/debug_marker",MarkerArray)
+        self.negate = False
+
+    @txros.util.cancellableInlineCallbacks
```
Like you did for the other ones, you should just return the defer instead of yielding on it inside of this function

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+        assert self.gate_3.found, "Not found"
+        self.gate_3_pos = navigator_tools.rosmsg_to_numpy(self.gate_3.objects[0].position)[:2]
+
+    def get_observation_poses(self):
```
I know it was written at the lake, but comments! (after the fact atleast)

mattlangford replied:
There are probably some issues here depending on what side of the gate's we're on. How are you handling that?

Also alot of this could be replaced by pose_editor commands (pretty much all the math since I know you're a library guy). *(see below)*

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+          yield self.navigator.mission_params["acoustic_pinger_active"].set("WHITE-GREEN")
+
+    def _get_perp(self,g1,g3):
+        assert len(g1) == 2
```
PEP in the function call.
Also, I wouldn't throw an assertion here, just do `g1 = np.array([g1])[:2]` which is much more functional (since you could  pass in a non numpy array also).

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+        g_perp = g_perp /  np.linalg.norm(g_perp)
+        self.g_perp = g_perp
+
+    def _thru_gate_poses(self, gatees, index):
```
`gatees` isn't used

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+
+    @txros.util.cancellableInlineCallbacks
+    def run(self):
+        print "Starting pinger"
```
You should use `fprint`, the colors are cool

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+    def get_observation_poses(self):
+        self._get_perp(self.gate_1_pos, self.gate_3_pos)
+
+        pose = yield self.navigator.tx_pose()
```
With `@property` decorators (like `navigator.tx_pose`'s) you don't need the `( )` at the end (it's treated like a class function), and it should throw an error if you do - but I think since that function is double decorated, nothing bad happens.

mattlangford commented on perception/navigator_vision/nodes/pinger_finder.py:
```diff
@@ -0,0 +1,169 @@
+      return
+    print "Got processed ping message:\n{}".format(processed_ping)
+    if abs(processed_ping.freq - target_freq) < freq_tol and processed_ping.amplitude > min_amp:
+        print "\x1b[34mTrustworthy pinger heading\x1b[0m"
```
*Cough* `fprint`

mattlangford commented on perception/navigator_vision/nodes/pinger_finder.py:
```diff
@@ -0,0 +1,169 @@
+    if abs(processed_ping.freq - target_freq) < freq_tol and processed_ping.amplitude > min_amp:
+        print "\x1b[34mTrustworthy pinger heading\x1b[0m"
+        trans, rot = None, None
+        try:
```
Probably should only enclose the TF stuff in the `try`. Also you may want to print out some kind of error notification so that the error message was caused by an error.

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+        pose = yield self.navigator.tx_pose()
+        self.dist_sign_pos = np.linalg.norm(pose[:2] - (self.gate_1_pos + self.OBSERVE_DISTANCE_METERS * self.g_perp) )
+        self.dist_sign_neg = np.linalg.norm(pose[:2] - (self.gate_1_pos - self.OBSERVE_DISTANCE_METERS * self.g_perp) )
+        if np.argmin( np.array( [self.dist_sign_pos,self.dist_sign_neg] ) ) == 1:
```
Why do you put spaces between some stuff like this?

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+    def get_objects(self):
+        self.gate_1 = yield self.navigator.database_query("Gate_1")
+        assert self.gate_1.found, "Not found"
+        self.gate_1_pos = navigator_tools.rosmsg_to_numpy(self.gate_1.objects[0].position)[:2]
```
You should really deal with all the points in 3d so you don't have to do a lot of appends later.

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+        self.gate_thru_poses  = self._thru_gate_poses( (self.gate_1_pos,self.gate_2_pos,self.gate_3_pos), self.gate_index)
+        self.gate_thru_markers()
+        for p in self.gate_thru_poses:
+            yield self.navigator.move.set_position(p).go(initial_plan_time=5)
```
You're really not using the pose_editor like to it's fullest extent.
```python
gate_pose = self.navigator.move.set_position(self.gate_1_pos).set_orientation(self.through_rot)
yield gate_pose.backward(10).go()
yield gate_position.forward(10).go(...)
```

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+          yield self.start_listen()
+          print "Listening at ",p
+          yield self.navigator.nh.sleep(self.LISTEN_TIME)
+        yield self.stop_listen()
```
Why have a `stop_listen` here?

ironmig replied:
Makes sure the pinger perception stops after we went to all observation points

mattlangford replied:
Ah yes, I see now. I must've thought the for loop was in a different order

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+        res = yield self.pinger_client(FindPingerRequest())
+        pinger_pose = res.pinger_position
+        self.pinger_pose = navigator_tools.rosmsg_to_numpy(pinger_pose)[:2]
+        self.distance1 = np.linalg.norm(self.pinger_pose - self.gate_1_pos) 
```
These distances could be computed in one fell swoop if the gate positions were a numpy list as well

mattlangford commented on mission_control/navigator_missions/nav_missions/pinger.py:
```diff
@@ -0,0 +1,198 @@
+        return (np.append(pose1,0), np.append(pose3,0))
+
+    @txros.util.cancellableInlineCallbacks
+    def observation_markers(self):
```
What are these two marker functions for?

mattlangford replied:
Lqrrt provides a goal pose that is visualizable whenever you set a way point.

ironmig commented:
I took out the actual pinger perception so David can PR it when he's ready (in whatever package he wants). Addressed most of the comments for the mission. The mission picks which side of the gate to observe on based on which is closer to current pose.  So far picking which color gate we went through is not implemented.

mattlangford commented:
I'll approve, but I recommend making the changes I had suggested in the future.
